tiny_dnn/core/kernels/fully_connected_op_internal.h

```diff
@@ -12,59 +12,87 @@
                                         const fully_params &params,
                                         const bool layer_parallelize) {
-  for_i(layer_parallelize, in_data.size(), [&](int sample) {
-    const vec_t &in = in_data[sample];
-    vec_t &out      = out_data[sample];
-
+  for_i(layer_parallelize, in_data.shape()[0], [&](int sample) {
     for (serial_size_t i = 0; i < params.out_size_; i++) {
-      out[i] = float_t{0};
+      out_data.host_at(sample, i) = float_t{0};
       for (serial_size_t c = 0; c < params.in_size_; c++) {
-        out[i] += W[c * params.out_size_ + i] * in[c];
+        out_data.host_at(sample, i) +=
+          weights.host_at(0, c * params.out_size_ + i) *
+          in_data.host_at(sample, c);
       }
 
       if (params.has_bias_) {
-        out[i] += bias[i];
+        out_data.host_at(sample, i) += bias.host_at(0, i);
       }
     }
   });
 }
 
-inline void fully_connected_op_internal(const tensor_t &prev_out,
-                                        const vec_t &W,
-                                        tensor_t &dW,
-                                        tensor_t &db,
-                                        tensor_t &curr_delta,
-                                        tensor_t &prev_delta,
+/**
+ * Backward propogation for fully connected layer with internal backend
+ * @param prev_out
+ * @param weigths
+ * @param weights_grads
+ * @param bias_grads
+ * @param curr_delta
+ * @param prev_delta
+ * @param params
+ * @param layer_parallelize
+ */
+template <typename S1,
+          typename S2,
+          typename S3,
+          typename S4,
+          typename S5,
+          typename S6>
+inline void fully_connected_op_internal(const Tensor<float_t, S1> &prev_out,
+                                        const Tensor<float_t, S2> &weigths,
+                                        Tensor<float_t, S3> &weights_grads,
+                                        Tensor<float_t, S4> &bias_grads,
+                                        Tensor<float_t, S5> &curr_delta,
+                                        Tensor<float_t, S6> &prev_delta,
                                         const fully_params &params,
                                         const bool layer_parallelize) {
-  for (serial_size_t sample = 0; sample < prev_out.size(); sample++) {
+  for (serial_size_t sample = 0; sample < prev_out.shape()[0]; sample++) {
     for (serial_size_t c = 0; c < params.in_size_; c++) {
       // propagate delta to previous layer
       // prev_delta[c] += current_delta[r] * W_[c * out_size_ + r]
-      prev_delta[sample][c] += vectorize::dot(
-        &curr_delta[sample][0], &W[c * params.out_size_], params.out_size_);
+      prev_delta.host_at(sample, c) += vectorize::dot(
+        curr_delta.host_iter(sample, 0),
+        weigths.host_iter(0, c * params.out_size_), params.out_size_);
     }
 
     for_(layer_parallelize, 0, size_t(params.out_size_),
```

Comment:
params structure doesn't make sense now since the info is encoded to the tensor itself